### PR TITLE
Querier: introduced `/api/v1/user_limits` endpoint.

### PR DESCRIPTION
**Pull request by wilfriedroset**
#### What this PR does

As a user, from an external point of view, I cannot know the limits applied to my tenant and how close am I to reach them. This route exposes realtime limits for the authenticated tenant, in JSON format. Along with `/api/v1/user_stats`, `<prometheus-http-prefix>/api/v1/cardinality/label_names` and `<prometheus-http-prefix>/api/v1/cardinality/label_values` I can better understand my usage and be proactive.

Alternatively, instead of exposing *all* limits struct we could create a dedicated struct `UserLimitsResponse` which would be a subset of the former with only what we would like to expose.

#### Which issue(s) this PR fixes or relates to

Relates #2341

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pstibrany** [comment]
> Alternatively, instead of exposing _all_ limits struct we could create a dedicated struct `UserLimitsResponse` which would be a subset of the former with only what we would like to expose.

We already have this, it's called `overrides-exporter`.  If you enable this module in Mimir, you will get `cortex_limits_overrides` metric for each user from runtime config, and also `cortex_limits_defaults` with default values.

Does that work for you? If there are missing limits that you need, perhaps we can just add those to overrides-exporter?

**Comment by wilfriedroset** [comment]
My goal is to expose this information realtime like `/api/v1/user_stats` instead of gathering (and persisting) those high cardinality series. They do not change *often* and I reckon exposing this information would be a good addition to the routes mentioned above.

More over, if the mimir admin enable and scrape those series they will be written in hist tenant and not end users tenant. As such, the admin will have the information which is great but not the end users.

I can rework the PR to expose only the information exposed by overrides-exporter. WDYT?

**Comment by pstibrany** [comment]
> My goal is to expose this information realtime like `/api/v1/user_stats` instead of gathering (and persisting) those high cardinality series. They do not change _often_ and I reckon exposing this information would be a good addition to the routes mentioned above.
> 
> More over, if the mimir admin enable and scrape those series they will be written in hist tenant and not end users tenant. As such, the admin will have the information which is great but not the end users.

I see, thanks for explanation. I'm not sure if distributor is the right component for this. Can we make the handler directly in `util/validation` package, and make it use `TenantLimits` interface, similar to `OverridesExporter`? We can then register it from `RegisterRuntimeConfig` (https://github.com/grafana/mimir/blob/b3bc0ab202bdee2ea145b5aade0fb137cb41a471/pkg/api/api.go#L230), so it will be part of any component using runtime config. WDYT?

> I can rework the PR to expose only the information exposed by overrides-exporter. WDYT?

Yes, I think that would be better. Many limits are low-level and user doesn't really need them.

Thank you!

File: pkg/util/validation/limits.go [hunk 1]
```diff
@@ -787,3 +790,59 @@ func MaxDurationPerTenant(tenantIDs []string, f func(string) time.Duration) time
 	}
 	return result
 }
+
+type userLimitsResponse struct {
```
Comment (wilfriedroset) [inline pkg/util/validation/limits.go:794]: WDYT about exposing `CompactorBlocksRetentionPeriod` as well?

With this information a tenant will be able to know for how long its data will be keep. I would say this should the retention period in seconds (int64). We have the case of unlimited retention `CompactorBlocksRetentionPeriod=0`, we could either return `0` as it is and point to the documentation where it is explained that `0` means unlimited or we can return `math.MaxInt64`. The former seems better.

Example of use of this information:
* In Grafafa with a `JSON API` data source type, I can GET this information and render it as a single stat while choosing the unit `Time / duration (s)`

Comment (pstibrany) [reply pkg/util/validation/limits.go:794]: I think exposing `CompactorBlocksRetentionPeriod` is fine, with 0 meaning "unlimited". It's consistent with config option.

Would it make sense to move this handler and struct to a separate file in this package?

File: pkg/util/validation/limits.go [hunk 2]
```diff
@@ -787,3 +790,59 @@ func MaxDurationPerTenant(tenantIDs []string, f func(string) time.Duration) time
...
+	MaxFetchedSeriesPerQuery     float64 `json:"max_fetched_series_per_query"`
+	MaxFetchedChunkBytesPerQuery float64 `json:"max_fetched_chunk_bytes_per_query"`
+
+	// Ruler limits
+	RulerMaxRulesPerRuleGroup   float64 `json:"ruler_max_rules_per_rule_group"`
+	RulerMaxRuleGroupsPerTenant float64 `json:"ruler_max_rule_groups_per_tenant"`
+}
+
+// UserLimitsHandler handles user limits.
+func UserLimitsHandler(defaultLimits Limits, tenantLimits TenantLimits) http.HandlerFunc {
```
Comment (pstibrany) [inline pkg/util/validation/limits.go:813]: Can we add a test for this handler?

File: pkg/util/validation/limits.go [hunk 3]
```diff
@@ -787,3 +790,59 @@ func MaxDurationPerTenant(tenantIDs []string, f func(string) time.Duration) time
...
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (pstibrany) [inline pkg/util/validation/limits.go:819]: This should return `http.StatusUnauthorized`:

```
		http.Error(w, err.Error(), http.StatusUnauthorized)
```

File: pkg/util/validation/limits.go [hunk 4]
```diff
@@ -787,3 +790,59 @@ func MaxDurationPerTenant(tenantIDs []string, f func(string) time.Duration) time
...
 	return result
 }
+
+type userLimitsResponse struct {
+	// Write path limits
+	IngestionRate             float64 `json:"ingestion_rate"`
+	IngestionBurstSize        float64 `json:"ingestion_burst_size"`
+	MaxGlobalSeriesPerUser    float64 `json:"max_global_series_per_user"`
+	MaxGlobalSeriesPerMetric  float64 `json:"max_global_series_per_metric"`
+	MaxGlobalExemplarsPerUser float64 `json:"max_global_exemplars_per_user"`
```
Comment (pstibrany) [inline pkg/util/validation/limits.go:800]: Many of these are `int`, not float. Why use float64 for everything? I'd suggest to keep the type same as original limit. (Same for read path and ruler limits)

File: pkg/util/validation/user_limits_handler.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	"net/http"
+	"time"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+type UserLimitsResponse struct {
+	CompactorBlocksRetentionPeriod float64 `json:"compactor_blocks_retention_period_second"` // suffix with second to make it explicit the value is in second
```
Comment (pstibrany) [inline pkg/util/validation/user_limits_handler.go:18]: In seconds this value will be quite big, and most people configure this in days/months. If we export this value in days, then values like `7.25` still make sense (= 7h4h). (If you prefer seconds, then let's use `int64` instead).

File: pkg/util/validation/user_limits_handler.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	RulerMaxRulesPerRuleGroup   int `json:"ruler_max_rules_per_rule_group"`
+	RulerMaxRuleGroupsPerTenant int `json:"ruler_max_rule_groups_per_tenant"`
+}
+
+// UserLimitsHandler handles user limits.
+func UserLimitsHandler(defaultLimits Limits, tenantLimits TenantLimits) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		userID, err := tenant.TenantID(ctx)
```
Comment (pstibrany) [inline pkg/util/validation/user_limits_handler.go:42]: We don't need `ctx` elsewhere:

```suggestion
		userID, err := tenant.TenantID(r.Context())
```

File: pkg/util/validation/user_limits_handler.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+			MaxGlobalSeriesPerMetric:  userLimits.MaxGlobalSeriesPerMetric,
+			MaxGlobalExemplarsPerUser: userLimits.MaxGlobalExemplarsPerUser,
+
+			// Read path limits
+			MaxChunksPerQuery:            userLimits.MaxChunksPerQuery,
+			MaxFetchedSeriesPerQuery:     userLimits.MaxFetchedSeriesPerQuery,
+			MaxFetchedChunkBytesPerQuery: userLimits.MaxFetchedChunkBytesPerQuery,
+			// Ruler limits
+
+			RulerMaxRulesPerRuleGroup:   userLimits.RulerMaxRulesPerRuleGroup,
```
Comment (pstibrany) [inline pkg/util/validation/user_limits_handler.go:69]:
```suggestion

			// Ruler limits
```

File: pkg/util/validation/user_limits_handler.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/pkg/util/validation/user_limits_exposition.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Cortex Authors.
+
```
Comment (pstibrany) [inline pkg/util/validation/user_limits_handler.go:5]: We only need to include these "Provenance" headers if the file was copied from Cortex. That's not the case here, so we can remove all Provenance headers here.

Same in `user_limits_handler_test.go`.

File: CHANGELOG.md [hunk 9]
```diff
@@ -11,6 +11,7 @@
 * [CHANGE] Query-frontend: CLI flag `-query-frontend.align-querier-with-step` has been deprecated. Please use `-query-frontend.align-queries-with-step` instead. #2840
 * [FEATURE] Introduced an experimental anonymous usage statistics tracking (disabled by default), to help Mimir maintainers make better decisions to support the open source community. The tracking system anonymously collects non-sensitive, non-personally identifiable information about the running Mimir cluster, and is disabled by default. #2643 #2662 #2685 #2732 #2733 #2735
 * [FEATURE] Introduced an experimental deployment mode called read-write and running a fully featured Mimir cluster with three components: write, read and backend. The read-write deployment mode is a trade-off between the monolithic mode (only one component, no isolation) and the microservices mode (many components, high isolation). #2754 #2838
+* [FEATURE] Introduced `/api/v1/user_limits` endpoint. This route exposes realtime limits for the authenticated tenant, in JSON format. #2864
```
Comment (pstibrany) [inline CHANGELOG.md:14]: * [FEATURE] Introduced `/api/v1/user_limits` endpoint exposed by all components that load runtime configuration. This endpoint exposes realtime limits for the authenticated tenant, in JSON format. #2864

**Review by pstibrany — COMMENTED** [review]
Thank you. Can you please also document the new endpoint in `docs/sources/operators-guide/reference-http-api/index.md` file?

File: pkg/util/validation/user_limits_handler.go (hunk 5 above)
Comment (wilfriedroset) [reply pkg/util/validation/user_limits_handler.go]: I would prefer seconds. With this type, the data is directly usable in Grafana

File: pkg/util/validation/user_limits_handler.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	"net/http"
+	"time"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+type UserLimitsResponse struct {
+	CompactorBlocksRetentionPeriod int64 `json:"compactor_blocks_retention_period_second"` // suffix with second to make it explicit the value is in second
```
Comment (pstibrany) [inline pkg/util/validation/user_limits_handler.go:13]:
```suggestion
	CompactorBlocksRetentionPeriod int64 `json:"compactor_blocks_retention_period_seconds"` // suffix with second to make it explicit the value is in seconds
```

File: pkg/util/validation/user_limits_handler_test.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+)
+
+func TestUserLimitsHandler(t *testing.T) {
+	var d model.Duration
+	_ = d.Set("1d") // don't need to check that 1d is a correct value
+
+	defaults := Limits{
+		IngestionRate:                  100,
+		IngestionBurstSize:             10,
+		CompactorBlocksRetentionPeriod: d, // verify this is converted to second as float64
```
Comment (pstibrany) [inline pkg/util/validation/user_limits_handler_test.go:22]: Nit: Comment is no longer correct.

File: pkg/util/validation/user_limits_handler.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
+package validation
```
Comment (pstibrany) [inline pkg/util/validation/user_limits_handler.go:1]: Apologies for not being clear... we still need the License-Identifier header, only Provenance headers should have been removed. (Same in test file).

```suggestion
// SPDX-License-Identifier: AGPL-3.0-only

package validation
```

**Review by pstibrany — APPROVED** [review]
Thank you for your work on this!

**Pull request merged**